src/javascript/app_2/Stores/Modules/Portfolio/portfolio_store.js

```diff
@@ -30,71 +31,77 @@ export default class PortfolioStore extends BaseStore {
     @action.bound
     portfolioHandler(response) {
         this.is_loading = false;
-        this.updatePortfolio(response);
+        if ('error' in response) {
+            this.error = response.error.message;
+            return;
+        }
+        this.error = '';
+        if (response.portfolio.contracts) {
+            this.data = response.portfolio.contracts
+                .map(pos => formatPortfolioPosition(pos))
+                .sort((pos1, pos2) => pos2.reference - pos1.reference); // new contracts first
+        }
     };
 
     @action.bound
     transactionHandler(response) {
         if ('error' in response) {
             this.error = response.error.message;
         }
-        WS.portfolio().then((res) => this.updatePortfolio(res));
-        // subscribe to new contracts:
-        WS.subscribeProposalOpenContract(null, this.proposalOpenContractHandler, false);
+        if (!response.transaction) return;
+        const { contract_id, action: act } = response.transaction;
+
+        if (act === 'buy') {
+            WS.portfolio().then((res) => {
+                const new_pos = res.portfolio.contracts.find(pos => +pos.contract_id === +contract_id);
+                if (!new_pos) return;
+                this.pushNewPosition(new_pos);
+            });
+            // subscribe to new contract:
+            WS.subscribeProposalOpenContract(contract_id, this.proposalOpenContractHandler, false);
```

Comment:
Subscribing to a particular `contract_id` causes a bug here, unless it handles forgetting it on contract expiry.
2nd option: It can be handled in Subscription Manager by removing the subscription when it receives a response without `id` (and not the first response for some calls!)

Reply:
@binary-hamedanchi by bug, you mean that expired contracts remain in `subscriptions`?

Reply:
@binary-hamedanchi i have gone with an option 2 :)